Introduce impl TryFrom for Number that succeeds iff the value is within the safe range

Ekleog opened the pull request:
As a counterpoint, raised by MaulingMonkey on discord: it could also make sense to round the number, so maybe TryFrom<u64> & co should not be implemented for Number?

OTOH, I personally think the added convenience is worth it, and by implementing TryFrom rather than From we make it explicit that the conversion is fallible, which should make it obvious that the behavior is "error out if the number is outside the safe range". Especially considering that `Number` in JavaScript is often used to describe integers in the safe range, probably much more often than it is actually used to describe floating point numbers.

A design question I had while writing this: should the `Error` type be just the passed-in value? It seems to be the way other `TryFrom` implementations in `js-sys` are currently done, so I went ahead with the same idea here.

daxpedda commented on crates/js-sys/src/lib.rs:
```diff
@@ -2793,6 +2793,24 @@ macro_rules! number_from {
+
+            #[inline]
+            fn try_from(x: $x) -> Result<Number, Self::Error> {
+                if x as f64 >= Number::MIN_SAFE_INTEGER && x as f64 <= Number::MAX_SAFE_INTEGER {
+                    Ok(Number::unchecked_from_js(JsValue::from(x as f64)))
```
```suggestion
                let x_f64 = x as f64;

                if x_f64 >= Number::MIN_SAFE_INTEGER && x_f64 <= Number::MAX_SAFE_INTEGER {
                    Ok(Number::unchecked_from_js(JsValue::from(x_f64)))
```

daxpedda commented on crates/js-sys/tests/wasm/Number.rs:
```diff
@@ -1,4 +1,7 @@
-use std::f64::{INFINITY, NAN};
+use std::{
+    convert::TryFrom,
+    f64::{INFINITY, NAN},
```
```suggestion
```
I assume this was from an earlier iteration?

Ekleog replied:
I'm not sure what you mean here? These imports were already present before my changes, and do seem to be required

daxpedda replied:
Apologies, I was sleep walking apparently :sweat_smile:!

daxpedda reviewed (changes requested):
> A design question I had while writing this: should the Error type be just the passed-in value? It seems to be the way other TryFrom implementations in js-sys are currently done, so I went ahead with the same idea here.

Where did you see that in `js-sys`?

---

I think I'm also in favor of doing the conversion like this, but FWIW alternatively we could make it infallible by using -/+infinity when the number is not in range.

Please add a changelog entry as well.

Liamolucko commented:
> > A design question I had while writing this: should the Error type be just the passed-in value? It seems to be the way other TryFrom implementations in js-sys are currently done, so I went ahead with the same idea here.
> 
> Where did you see that in `js-sys`?

The `TryFrom<BigInt>` impls for `i64`, `u64`, `i128` and `u128` do that, as well as all the `TryFrom` implementations in `wasm-bindgen`.

The argument against doing the same here would be that all those impls convert from `JsValue` (or an imported JS type) to a standard Rust type; you might want to get the `JsValue` back on failure to avoid having to needlessly clone it. In this case, we're converting from an integer to a `JsValue`, so there isn't as much need to get the original value back on failure when the input's `Copy` anyway.

That would line up with what `std` does: looking at all of `std`'s implementations of `TryFrom<T>`, it seems like whenever `T` can't be cheaply cloned it gives back the `T` on failure, otherwise it uses a real error type. For example `impl TryFrom<Vec<T>> for [T; N]` has `type Error = Vec<T>`, whereas `impl TryFrom<&[T]> for [T; N]` has `type Error = FromSliceError`. I don't actually know if this is an official policy or anything but it seems to be the case.

The closest analogy to `JsValue` would be `impl TryFrom<Arc<[T]>> for Arc<[T; N]>`, since both are cheap but not free to clone: and it gives back the `Arc` on failure. So I guess using a proper error type would make more sense.

Ekleog commented:
Thank you for your feedback! I have:
- hoisted `x_f64` as per the suggestion
- added a changelog entry
- switched the code to use a `TryFromIntError` type, copy-pasted from the libstd, the only change being that I didn't import `Error` globally. We apparently cannot construct the libstd's variant easily, and doing eg. `u32::try_from(u64::MAX)?` to return it would look like too much of a hack for me — let me know if you'd have preferred that approach

Hopefully this is good to go! :)

daxpedda commented on CHANGELOG.md:
```diff
@@ -11,6 +11,8 @@
 * Add bindings for `CanvasState.reset()`, affecting `CanvasRenderingContext2D` and `OffscreenCanvasRenderingContext2D`.
   [#3844](https://github.com/rustwasm/wasm-bindgen/pull/3844)
 
+* Add `TryFrom` implementations for `Number`, that allow losslessly converting from 64- and 128-bits numbers
```
```suggestion
* Add `TryFrom` implementations for `Number`, that allow losslessly converting from 64- and 128-bits numbers.
  [#3847](https://github.com/rustwasm/wasm-bindgen/pull/3847)
```

daxpedda commented on crates/js-sys/src/lib.rs:
```diff
@@ -2793,6 +2793,45 @@ macro_rules! number_from {
+            fn try_from(x: $x) -> Result<Number, Self::Error> {
+                let x_f64 = x as f64;
+                if x_f64 >= Number::MIN_SAFE_INTEGER && x_f64 <= Number::MAX_SAFE_INTEGER {
+                    Ok(Number::unchecked_from_js(JsValue::from(x_f64)))
```
```suggestion
                    Ok(Number::from(x_f64))
```
Just figured this could work as well.

daxpedda reviewed (changes requested):
> * switched the code to use a `TryFromIntError` type, copy-pasted from the libstd, the only change being that I didn't import `Error` globally. We apparently cannot construct the libstd's variant easily, and doing eg. `u32::try_from(u64::MAX)?` to return it would look like too much of a hack for me — let me know if you'd have preferred that approach

Yeah, my main concern with using Stds `TryFromIntError` type is that we aren't really in control of what it displays exactly so it could become confusing in the future. Though I'm not happy with adding a dedicated error type for this I don't have a good alternative suggestion.

So except the couple of nits its LGTM.

daxpedda commented on crates/js-sys/src/lib.rs:
```diff
@@ -2793,6 +2793,37 @@ macro_rules! number_from {
 
+/// The error type returned when a checked integral type conversion fails.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct TryFromIntError(pub(crate) ());
```
```suggestion
pub struct TryFromIntError(());
```

daxpedda reviewed (approved):
No, I will go ahead and fix the CI first.

Ekleog commented:
Awesome, thank you! I just handled your review comment and merged master in, hopefully this is now good to go :)

daxpedda commented:
@Liamolucko would appreciate your approval here as well.

Liamolucko commented on crates/js-sys/src/lib.rs:
```diff
@@ -2793,6 +2793,37 @@ macro_rules! number_from {
+macro_rules! number_try_from {
+    ($($x:ident)*) => ($(
+        impl TryFrom<$x> for Number {
+            type Error = $x;
```
You never actually updated this to use `TryFromIntError`.

Ekleog replied:
Oh. Now I feel stupid ._. This is now fixed, thank you!
